failed to init with cmake 3.16.0

I have no issue with cmake 3.15.5 on Ubuntu 18.04, but it failed to init when I update to the newest cmake version.

> Hi @cvluca , are you able to provide a build log?

@kent-mcleod  There is no cmake log being generated. This is what the error looks like.

```
$ ./init --tut hello-world
loading initial cache file ../projects/sel4-tutorials/settings.cmake
-- Found capdl: /home/luca/sel4/tutorial/projects/capdl
-- Found camkes-tool: /home/luca/sel4/tutorial/projects/sel4-tutorials
CMake Error: The source directory "/home/luca/sel4/tutorial/hello-world" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
Traceback (most recent call last):
  File "./init", line 96, in <module>
  File "./init", line 85, in main
  File "/home/luca/sel4/tutorial/projects/sel4-tutorials/common.py", line 106, in init_directories
    return _init_build_directory(config, initialised, build_directory, tute_directory, output, config_dict=config_dict)
  File "/home/luca/sel4/tutorial/projects/sel4-tutorials/common.py", line 74, in _init_build_directory
    return sh.cmake(args + [tute_directory], _cwd=directory, _out=output, _err=output)
  File "/home/luca/.local/lib/python3.6/site-packages/sh.py", line 1427, in __call__
    return RunningCommand(cmd, call_args, stdin, stdout, stderr)
  File "/home/luca/.local/lib/python3.6/site-packages/sh.py", line 774, in __init__
    self.wait()
  File "/home/luca/.local/lib/python3.6/site-packages/sh.py", line 792, in wait
    self.handle_command_exit_code(exit_code)
  File "/home/luca/.local/lib/python3.6/site-packages/sh.py", line 815, in handle_command_exit_code
    raise exc
sh.ErrorReturnCode_1:

  RAN: /usr/local/bin/cmake -G Ninja -DTUT_BOARD=pc -DTUT_ARCH=x86_64 -DTUTORIAL_DIR=hello-world -C ../projects/sel4-tutorials/settings.cmake /home/luca/sel4/tutorial/hello-world

  STDOUT:


  STDERR:
```

--- a/Findsel4-tutorials.cmake
+++ b/Findsel4-tutorials.cmake
@@ -20,7 +20,7 @@ macro(sel4_tutorials_regenerate_tutorial tutorial_dir)
     # generate tutorial sources into directory
 
     GenerateTutorial(${tutorial_dir})
-    if(${CMAKE_CURRENT_LIST_DIR} STREQUAL ${CMAKE_CURRENT_SOURCE_DIR})
+    if("${CMAKE_CURRENT_LIST_FILE}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}/CMakeLists.txt")
         get_property(tute_hack GLOBAL PROPERTY DONE_TUTE_HACK)
         if(NOT tute_hack)
             set_property(GLOBAL PROPERTY DONE_TUTE_HACK TRUE)

--- a/Findsel4-tutorials.cmake
+++ b/Findsel4-tutorials.cmake
@@ -20,16 +20,12 @@ macro(sel4_tutorials_regenerate_tutorial tutorial_dir)
     # generate tutorial sources into directory
 
     GenerateTutorial(${tutorial_dir})
-    # Add the tutorial directory the same as any other project
-    # add_subdirectory(${CMAKE_SOURCE_DIR}/${TUTORIAL_DIR} ${CMAKE_BINARY_DIR}/${TUTORIAL_DIR})
-    # message("${CMAKE_CURRENT_LIST_DIR} ${CMAKE_CURRENT_SOURCE_DIR}")
     if(${CMAKE_CURRENT_LIST_DIR} STREQUAL ${CMAKE_CURRENT_SOURCE_DIR})
         get_property(tute_hack GLOBAL PROPERTY DONE_TUTE_HACK)
         if(NOT tute_hack)
             set_property(GLOBAL PROPERTY DONE_TUTE_HACK TRUE)
             # We are in the main project phase and regenerating the tutorial
             # may have updated the file calling us.  So we do some magic...
-            # message("err ${CMAKE_CURRENT_SOURCE_DIR}")
             include(${CMAKE_CURRENT_SOURCE_DIR}/CMakeLists.txt)
             return()
         endif()

--- a/cmake/helpers.cmake
+++ b/cmake/helpers.cmake
@@ -144,7 +144,19 @@ endfunction()
 function(GenerateTutorial full_path)
     get_filename_component(dir ${full_path} NAME)
     get_filename_component(base_dir ${full_path} DIRECTORY)
-    include_guard(GLOBAL)
+
+    # Implement include_guard() functionality within a function that
+    # can only be called once. Using include_guard() is considered dangerous
+    # within functions as it will be affected by other include_guard() calls
+    # within functions within the same file.
+    get_property(gen_tutorial GLOBAL PROPERTY GenerateTutorialONCE)
+    if(NOT gen_tutorial)
+        set_property(GLOBAL PROPERTY GenerateTutorialONCE TRUE)
+    else()
+        # If already generated the tutorial in this run we don't need to do so again.
+        return()
+    endif()
+
     if(EXISTS ${base_dir}/${dir}/.tute_config)
         set(output_dir ${CMAKE_CURRENT_BINARY_DIR}/.tutegen/${dir}/gen)
         set(old_output_dir ${CMAKE_CURRENT_BINARY_DIR}/.tutegen/${dir}/old)

--- a/cmake/helpers.cmake
+++ b/cmake/helpers.cmake
@@ -124,7 +124,7 @@ function(ExecuteGenerationProcess input_dir output_dir generated_files)
     # Set cmake to regenerate if any of the input files to the TUTE_COMMAND are updated
     file(READ "${input_files}" files)
     separate_arguments(file_list NATIVE_COMMAND ${files})
-    if(NOT ${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
+    if("${CMAKE_CURRENT_LIST_FILE}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}/CMakeLists.txt")
         set_property(
             DIRECTORY "${CMAKE_SOURCE_DIR}"
             APPEND

--- a/settings.cmake
+++ b/settings.cmake
@@ -10,9 +10,8 @@
 # @TAG(DATA61_BSD)
 #
 
-if(NOT ${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
-    include_guard(GLOBAL)
-endif()
+include_guard(GLOBAL)
+
 set(project_dir "${CMAKE_CURRENT_LIST_DIR}/../../")
 file(GLOB project_modules ${project_dir}/projects/*)
 list(

